Docs for new default PodTopologySpread functionality and gate (#21689)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md

### Cluster-level default constraints

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

> 

Reviewer: From `kubescheduler.config.k8s.io/v1beta1` I'm inferring that at least part of this is beta. What's the best way to explain to readers about whether or not this kind of setting is GA?

Author: In the sense that the configuration API is beta, then yes, this could be considered beta as well. @Huang-Wei do you agree?

Reviewer 2: I'm fine with either. Actually the code quality for this part is GA, but users can still disable (which is the default case) it by not setting the arg in scheduler configuration (v1beta1).

Author: I put the annotation back (but beta)

      args:
        defaultConstraints:
          - maxSkew: 1
            topologyKey: topology.kubernetes.io/zone
            whenUnsatisfiable: ScheduleAnyway
```

{{< note >}}
The score produced by default scheduling constraints might conflict with the
score produced by the
[`SelectorSpread` plugin](/docs/reference/scheduling/config/#scheduling-plugins).
It is recommended that you disable this plugin in the scheduling profile when
using default constraints for `PodTopologySpread`.
{{< /note >}}

#### Internal default constraints

> {{< feature-state for_k8s_version="v1.19" state="alpha" >}}

Reviewer 3: Can you update the [Feature gate page](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) as well? You need to do the following:

- Add the `DefaultPodTopologySpread` feature gate to the feature gate table
- Add a short description of `DefaultPodTopologySpread` to the list of feature gates (and link back to this page of docs!)

Author: Done

> When the feature gate `DefaultPodTopologySpreadPlugin` is enabled,

Reviewer 4: ```suggestion
When the feature gate `DefaultPodTopologySpread` is enabled,
```

Author: Oops. Thanks

Reviewer 2: didn't catch that...

> When the feature gate `DefaultPodTopologySpread` is enabled,
> kube-scheduler uses the following default topology constraints for the
> `PodTopologySpread` plugin configuration:

Reviewer 3: ```suggestion
When you enable the `DefaultPodTopologySpread`, the legacy `SelectorSpread` plugin is disabled.
kube-scheduler uses the following default topology constraints for the
`PodTopologySpread` plugin configuration:
```

Author: Done

```yaml
defaultConstraints:
  - maxSkew: 3
    topologyKey: "kubernetes.io/hostname"
    whenUnsatisfiable: ScheduleAnyway
  - maxSkew: 5
    topologyKey: "topology.kubernetes.io/zone"
    whenUnsatisfiable: ScheduleAnyway
```

> Also, the legacy `SelectorSpread` plugin, which provides an equivalent behavior,
> is disabled.

Reviewer 3: ```suggestion
```

See above – better to have this all in one chunk before the example block.

Author: Done

{{< note >}}

> If your nodes are not expected to have **both** `kubernetes.io/hostname` and
> `topology.kubernetes.io/zone` labels set, you should define your own constraints
> instead of using the Kubernetes default.

Reviewer 3: ```suggestion
If your nodes are not expected to have **both** `kubernetes.io/hostname` and
`topology.kubernetes.io/zone` labels set, define your own constraints
instead of using the Kubernetes defaults.
```

Author: Done

The `PodTopologySpread` plugin ignores from scoring the nodes that don't have

> both topologies.

Reviewer 3: Unsure what you're trying to say here. It's either:

> The `PodTopologySpread` plugin ignores scoring from the nodes that don't have
both topologies.

or

> The `PodTopologySpread` plugin does not score scoring the nodes that don't have
both topologies.

Either way this sentence isn't grammatically correct?

Author: See update

using default constraints for `PodTopologySpread`.
{{< /note >}}

> #### A kubernetes default

Reviewer: ```suggestion
#### Default topology constraints
```

Author: My concern is the parent section "Cluster-level default constraints"

Basically, this section is describing the "default of default". So, ideally it should be "Default cluster-level default constraints", but I'm not sure if it reads well.

WDYT?

{{< feature-state for_k8s_version="v1.19" state="alpha" >}}

> When the feature gate `DefaultPodTopologySpreadPlugin` is enabled, the following
> default constraints for `PodTopologySpreadPlugin` are applied:

Reviewer: Can you rewrite this in the active voice?
Eg:
```suggestion
When the feature gate `DefaultPodTopologySpreadPlugin` is enabled, the control
plane applies the following default topology constraints:
```

```yaml
defaultConstraints:
  - maxSkew: 3
    topologyKey: "kubernetes.io/hostname"
    whenUnsatisfiable: ScheduleAnyway
  - maxSkew: 5
    topologyKey: "topology.kubernetes.io/zone"
    whenUnsatisfiable: ScheduleAnyway
```

Also, the legacy `SelectorSpread` plugin is disabled.

{{< note >}}
If your nodes are not expected to have **both** `kubernetes.io/hostname` and

> `topology.kubernetes.io/zone` we recommend that you define your own constraints

Reviewer: _nit_, doesn't block a merge
To [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we), how about:
```suggestion
`topology.kubernetes.io/zone` keys set, you should define your own constraints
```
?

> instead of using the kubernetes default.

Reviewer: _nit_, doesn't block a merge
```suggestion
instead of using the Kubernetes default.
```

The `PodTopologySpread` plugin ignores all nodes that don't have both topologies

> from scoring.

Reviewer: Does this refer to the two topology keys `kubernetes.io/hostname` and `topology.kubernetes.io/zone`? I wasn't sure.

Author: Yes. But let me describe it more generally.

The `PodTopologySpread` plugin gives a positive score to nodes if and only if the node has all the topologies used in the spreading constraints. In the case of the default, this resolves to hostname and zone.
